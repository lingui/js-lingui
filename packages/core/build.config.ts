import { defineBuildConfig } from "unbuild"

export default defineBuildConfig({
  externals: ["@lingui/core"],
  failOnWarn: false,
})
