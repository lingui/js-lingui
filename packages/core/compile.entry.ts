/**
 * this is Rollup Entry
 */
export * from "./src/compile"
