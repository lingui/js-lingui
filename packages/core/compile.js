/**
 * This entry is for old runtimes which do not support `exports` field in package.json
 * https://github.com/facebook/metro/issues/670
 */
module.exports = require("./build/cjs/compile.js")
