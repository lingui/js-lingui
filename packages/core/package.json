{
  "name": "@lingui/core",
  "version": "2.1.2",
  "description": "I18n tools for javascript",
  "keywords": [
    "i18n",
    "i10n",
    "i9n",
    "internationalization",
    "localization",
    "translation"
  ],
  "repository": "lingui/js-lingui",
  "bugs": "https://github.com/lingui/js-lingui/issues",
  "license": "MIT",
  "author": {
    "name": "Tomáš Ehrlich",
    "email": "tomas.ehrlich@gmail.com"
  },
  "main": "index.js",
  "engines": {
    "node": ">=4.0"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.js",
    "dev.js",
    "cjs/"
  ],
  "dependencies": {
    "babel-runtime": "^6.26.0",
    "make-plural": "^4.1.1",
    "messageformat-parser": "^2.0.0"
  }
}
