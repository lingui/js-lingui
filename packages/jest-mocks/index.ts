export function getConsoleMockCalls({ mock }: jest.MockInstance<any, any>) {
  if (!mock.calls.length) return
  return mock.calls.map((call) => call[0]).join("\n")
}

export function mockConsole(
  testCase: (console: jest.Mocked<Console>) => any,
  mock = {}
) {
  function restoreConsole() {
    global.console = originalConsole
  }

  const originalConsole = global.console

  const defaults = {
    log: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
  }

  global.console = {
    ...defaults,
    ...mock,
  } as any

  let result
  try {
    result = testCase(global.console as jest.Mocked<Console>)
  } catch (e) {
    restoreConsole()
    throw e
  }

  if (result && typeof result.then === "function") {
    return result.then(restoreConsole).catch((e) => {
      restoreConsole()
      throw e
    })
  } else {
    restoreConsole()
    return result
  }
}

export function mockEnv(env, testCase) {
  const oldEnv = process.env.NODE_ENV
  process.env.NODE_ENV = env

  try {
    testCase()
  } catch (e) {
    process.env.NODE_ENV = oldEnv
    throw e
  }

  process.env.NODE_ENV = oldEnv
}
