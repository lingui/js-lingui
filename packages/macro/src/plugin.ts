import type { PluginObj, Visitor, PluginPass } from "@babel/core"
import type * as babelTypes from "@babel/types"
import MacroJSX from "./macroJsx"
import { NodePath } from "@babel/traverse"
import MacroJs from "./macroJs"
import { MACRO_PACKAGE } from "./constants"
import {
  type LinguiConfigNormalized,
  getConfig as loadConfig,
} from "@lingui/conf"
import { Program, Identifier } from "@babel/types"

let config: LinguiConfigNormalized

export type LinguiPluginOpts = {
  // explicitly set by CLI when running extraction process
  extract?: boolean
  linguiConfig?: LinguiConfigNormalized
}

function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

function reportUnsupportedSyntax(path: NodePath, e: Error) {
  const codeFrameError = path.buildCodeFrameError(
    `Unsupported macro usage. Please check the examples at https://lingui.dev/ref/macro#examples-of-js-macros.
 If you think this is a bug, fill in an issue at https://github.com/lingui/js-lingui/issues

 Error: ${e.message}`
  )

  // show stack trace where error originally happened
  codeFrameError.stack = e.stack
  throw codeFrameError
}

type LinguiSymbol = "Trans" | "useLingui" | "i18n"

export default function ({
  types: t,
}: {
  types: typeof babelTypes
}): PluginObj {
  function addImport(state: PluginPass, name: LinguiSymbol) {
    const path = state.get(
      "macroImport"
    ) as NodePath<babelTypes.ImportDeclaration>

    const config = state.get("linguiConfig") as LinguiConfigNormalized

    if (!state.get("has_import_" + name)) {
      state.set("has_import_" + name, true)
      const [moduleSource, importName] = config.runtimeConfigModule[name]

      const [newPath] = path.insertAfter(
        t.importDeclaration(
          [
            t.importSpecifier(
              getSymbolIdentifier(state, name),
              t.identifier(importName)
            ),
          ],
          t.stringLiteral(moduleSource)
        )
      )

      path.parentPath.scope.registerDeclaration(newPath)
    }

    return path.parentPath.scope.getBinding(
      getSymbolIdentifier(state, name).name
    )
  }

  function getMacroImports(path: NodePath<Program>) {
    return path.get("body").filter((statement) => {
      return (
        statement.isImportDeclaration() &&
        statement.get("source").node.value === MACRO_PACKAGE
      )
    })
  }

  function getSymbolIdentifier(
    state: PluginPass,
    name: LinguiSymbol
  ): Identifier {
    return state.get("linguiIdentifiers")[name]
  }

  return {
    name: "lingui-macro-plugin",
    visitor: {
      Program: {
        enter(path, state) {
          const macroImports = getMacroImports(path)

          if (!macroImports.length) {
            return
          }

          state.set("macroImport", macroImports[0])

          state.set(
            "linguiConfig",
            getConfig((state.opts as LinguiPluginOpts).linguiConfig)
          )

          state.set("linguiIdentifiers", {
            i18n: path.scope.generateUidIdentifier("i18n"),
            Trans: path.scope.generateUidIdentifier("Trans"),
            useLingui: path.scope.generateUidIdentifier("useLingui"),
          })

          path.traverse(
            {
              JSXElement(path, state) {
                const macro = new MacroJSX(
                  { types: t },
                  {
                    transImportName: getSymbolIdentifier(state, "Trans").name,
                    stripNonEssentialProps:
                      process.env.NODE_ENV == "production" &&
                      !(state.opts as LinguiPluginOpts).extract,
                  }
                )

                let newNode: false | babelTypes.Node

                try {
                  newNode = macro.replacePath(path)
                } catch (e) {
                  reportUnsupportedSyntax(path, e as Error)
                }

                if (newNode) {
                  const [newPath] = path.replaceWith(newNode)
                  addImport(state, "Trans").reference(newPath)
                }
              },

              "CallExpression|TaggedTemplateExpression"(
                path: NodePath<
                  | babelTypes.CallExpression
                  | babelTypes.TaggedTemplateExpression
                >,
                state: PluginPass
              ) {
                const macro = new MacroJs(
                  { types: t },
                  {
                    stripNonEssentialProps:
                      process.env.NODE_ENV == "production" &&
                      !(state.opts as LinguiPluginOpts).extract,
                    i18nImportName: getSymbolIdentifier(state, "i18n").name,
                    useLinguiImportName: getSymbolIdentifier(state, "useLingui")
                      .name,
                  }
                )
                let newNode: false | babelTypes.Node

                try {
                  newNode = macro.replacePath(path)
                } catch (e) {
                  reportUnsupportedSyntax(path, e as Error)
                }

                if (newNode) {
                  const [newPath] = path.replaceWith(newNode)

                  if (macro.needsUseLinguiImport) {
                    addImport(state, "useLingui").reference(newPath)
                  }

                  if (macro.needsI18nImport) {
                    addImport(state, "i18n").reference(newPath)
                  }
                }
              },
            } as Visitor<PluginPass>,
            state
          )
        },
        exit(path, state) {
          const macroImports = getMacroImports(path)
          macroImports.forEach((path) => path.remove())
        },
      },
    } as Visitor<PluginPass>,
  }
}
