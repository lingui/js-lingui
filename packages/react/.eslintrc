{
  "extends": ["../../.eslintrc", "plugin:react/recommended", "plugin:react-hooks/recommended"],
  "settings": {
    "react": {
      "version": "detect" // "detect" automatically picks the version of React that is installed
    }
  },
  "rules": {
    "react/prop-types": "off",
    "react/display-name": "off"
  }
}
