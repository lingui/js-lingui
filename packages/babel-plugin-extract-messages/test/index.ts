import path from "path"
import fs from "fs"
import { transform as babelTransform } from "@babel/core"
import plugin, { ExtractedMessage, ExtractPluginOpts } from "../src/index"
import { mockConsole } from "@lingui/jest-mocks"

const transform = (filename: string) => {
  const rootDir = path.join(__dirname, "fixtures")

  const filePath = path.join(rootDir, filename)
  const code = fs.readFileSync(filePath).toString()

  return transformCode(code, filePath, rootDir)
}

const transformCode = (
  code: string,
  filename = "test-case.js",
  rootDir = "."
) => {
  process.env.LINGUI_CONFIG = path.join(
    __dirname,
    "fixtures",
    "lingui.config.js"
  )
  const messages: ExtractedMessage[] = []

  try {
    const pluginOpts: ExtractPluginOpts = {
      onMessageExtracted(msg: ExtractedMessage) {
        const filename = path.relative(rootDir, msg.origin[0])
        messages.push({
          ...msg,
          origin: [filename, msg.origin[1]],
        })
      },
    }

    babelTransform(code, {
      configFile: false,
      filename,
      plugins: [
        "@babel/plugin-syntax-jsx",
        [
          "macros",
          {
            lingui: { extract: true },
            // macro plugin uses package `resolve` to find a path of macro file
            // this will not follow jest pathMapping and will resolve path from ./build
            // instead of ./src which makes testing & developing hard.
            // here we override resolve and provide correct path for testing
            resolvePath: (source: string) => require.resolve(source),
          },
        ],
        [plugin, pluginOpts],
      ],
    })
  } finally {
    process.env.LINGUI_CONFIG = null
  }

  return messages
}

function expectNoConsole(cb: () => void) {
  return mockConsole((console) => {
    cb()

    expect(console.warn).not.toBeCalled()
    expect(console.error).not.toBeCalled()
  })
}

describe("@lingui/babel-plugin-extract-messages", function () {
  it("should ignore files without lingui import", () => {
    expectNoConsole(() => {
      const messages = transform("without-lingui.js")
      expect(messages.length).toBe(0)
    })
  })

  it("should extract all messages from JSX files", () => {
    expectNoConsole(() => {
      const messages = transform("jsx-without-macros.js")
      expect(messages).toMatchSnapshot()
    })
  })

  describe("JSX", () => {
    it("Should not rise warning when translation from variable", () => {
      const code = `
import { Trans } from "@lingui/react";

<Trans id={message} />;
<Trans id={message.field} />;
      `
      expectNoConsole(() => {
        const messages = transformCode(code)
        expect(messages.length).toBe(0)
      })
    })

    it("Should log error when no ID provided", () => {
      const code = `
import { Trans } from "@lingui/react";

<Trans />;
<Trans message="Missing ID" />;
      `
      mockConsole((console) => {
        const messages = transformCode(code)

        expect(messages.length).toBe(0)
        expect(console.error).not.toBeCalled()
        expect(console.warn).toBeCalledWith(
          expect.stringContaining(`Missing message ID`)
        )
      })
    })
  })

  describe("CallExpression i18n._()", () => {
    it("should extract messages from i18n._ call expressions", () => {
      expectNoConsole(() => {
        const messages = transform("js-call-expression.js")
        expect(messages).toMatchSnapshot()
      })
    })

    it("should extract from member access expressions", () => {
      const code = `
      // member access
      ctx.i18n._("Message")
      
      // member access any depth
      ctx.req.i18n._("Message")
      `
      expectNoConsole(() => {
        const messages = transformCode(code)
        expect(messages.length).toBe(2)
      })
    })

    it("should not extract from random i18n members", () => {
      const code = `
      i18n.load("Message")
      `
      expectNoConsole(() => {
        const messages = transformCode(code)
        expect(messages.length).toBe(0)
      })
    })

    it("should not extract if disabled via annotation", () => {
      const code = `
      /* lingui-extract-ignore */
      i18n._("Message")
            
      /* lingui-extract-ignore */
      ctx.i18n._("Message")
      
       /* lingui-extract-ignore */
      ctx.req.i18n._("Message")
      `
      expectNoConsole(() => {
        const messages = transformCode(code)
        expect(messages.length).toBe(0)
      })
    })

    it("Should not rise warning when translation from variable", () => {
      const code = `
      i18n._(message);
      // member expression
      i18n._(foo.bar);
      // function call
      i18n._(getMessage());
      `

      expectNoConsole(() => {
        const messages = transformCode(code)
        expect(messages.length).toBe(0)
      })
    })

    it("Should throw error when not a string provided as comment", () => {
      const code = `const msg = i18n._('message.id', {}, {comment: variable})`

      return mockConsole((console) => {
        transformCode(code)

        expect(console.error).not.toBeCalled()
        expect(console.warn).toBeCalledWith(
          expect.stringContaining(
            "Only strings or template literals could be extracted."
          )
        )
      })
    })

    it("Should support extract id from TplLiteral and Concatenation", () => {
      const code = `
      const msg = i18n._(\`message.id\`);
      const msg2 = i18n._("second" + '.' + "id")
      `

      expectNoConsole(() => {
        const messages = transformCode(code)
        expect(messages[0]).toMatchObject({
          id: "message.id",
        })
        expect(messages[1]).toMatchObject({
          id: "second.id",
        })
      })
    })

    it("Should support string concatenation", () => {
      const code = `const msg = i18n._('message.id', {}, {comment: "first " + "second " + "third"})`

      expectNoConsole(() => {
        const messages = transformCode(code)
        expect(messages[0]).toMatchObject({
          comment: "first second third",
        })
      })
    })

    it("Should not double extract from call generated by macro and dont rise warnings", () => {
      const code = `import { i18n } from "@lingui/core";
    const msg =
      i18n._(/*i18n*/
        {
          id: "Hello {name}",
          values: {
            name: name,
          },
        });
      `
      expectNoConsole(() => {
        const messages = transformCode(code)
        expect(messages.length).toBe(1)
      })
    })
  })

  describe("StringLiteral", () => {
    it("Should extract from marked StringLiteral", () => {
      const code = `const t = /*i18n*/'Message'`

      expectNoConsole(() => {
        const messages = transformCode(code)

        expect(messages[0]).toMatchObject({
          id: "Message",
        })
      })
    })

    it("Should log error when empty StringLiteral marked for extraction", () => {
      const code = `const t = /*i18n*/''`

      return mockConsole((console) => {
        const messages = transformCode(code)

        expect(messages.length).toBe(0)
        expect(console.error).not.toBeCalled()
        expect(console.warn).toBeCalledWith(
          expect.stringContaining(`Empty StringLiteral`)
        )
      })
    })
  })

  describe("MessageDescriptor", () => {
    it("should extract messages from MessageDescriptors", () => {
      expectNoConsole(() => {
        const messages = transform("js-message-descriptor.js")
        expect(messages).toMatchSnapshot()
      })
    })

    it("Should extract id from TemplateLiteral", () => {
      const code = "const msg = /*i18n*/{id: `Message`}"

      expectNoConsole(() => {
        const messages = transformCode(code)

        expect(messages[0]).toMatchObject({
          id: "Message",
        })
      })
    })

    it("Should log error if TemplateLiteral in id has expressions", () => {
      const code = "const msg = /*i18n*/{id: `Hello ${name}`}"

      return mockConsole((console) => {
        const messages = transformCode(code)

        expect(messages.length).toBe(0)

        expect(console.error).not.toBeCalled()
        expect(console.warn).toBeCalledWith(
          expect.stringContaining(
            `Could not extract from template literal with expressions.`
          )
        )
      })
    })

    it("Should log error when no ID provided", () => {
      const code = "const msg = /*i18n*/ {message: `Hello ${name}`}"

      return mockConsole((console) => {
        const messages = transformCode(code)

        expect(messages.length).toBe(0)
        expect(console.error).not.toBeCalled()

        expect(console.warn).toBeCalledWith(
          expect.stringContaining(`Missing message ID`)
        )
      })
    })

    it("Should log error when not a string provided as ID", () => {
      const code = "const msg = /*i18n*/ {id: id}"

      return mockConsole((console) => {
        transformCode(code)

        expect(console.error).not.toBeCalled()
        expect(console.warn).toBeCalledWith(
          expect.stringContaining(
            "Only strings or template literals could be extracted."
          )
        )
      })
    })

    it("Should log error when not a string provided as comment", () => {
      const code = `const msg =  /*i18n*/ {id: "msg.id", comment: variable}`

      return mockConsole((console) => {
        transformCode(code)

        expect(console.error).not.toBeCalled()
        expect(console.warn).toBeCalledWith(
          expect.stringContaining(
            "Only strings or template literals could be extracted."
          )
        )
      })
    })

    it("Should support string concatenation in comment", () => {
      const code = `const msg =  /*i18n*/ {id: "msg.id", comment: "first " + "second " + "third"}`

      return expectNoConsole(() => {
        const messages = transformCode(code)
        expect(messages[0]).toMatchObject({
          comment: "first second third",
        })
      })
    })
  })

  it("should extract all messages from JSX files (macros)", () => {
    return expectNoConsole(() => {
      const messages = transform("jsx-with-macros.js")
      expect(messages).toMatchSnapshot()
    })
  })

  it("should extract Plural messages from JSX files when there's no Trans tag (integration)", () => {
    return expectNoConsole(() => {
      const messages = transform("jsx-without-trans.js")
      expect(messages).toMatchSnapshot()
    })
  })

  it("should extract all messages from JS files (macros)", () => {
    return expectNoConsole(() => {
      const messages = transform("js-with-macros.js")
      expect(messages).toMatchSnapshot()
    })
  })
})
