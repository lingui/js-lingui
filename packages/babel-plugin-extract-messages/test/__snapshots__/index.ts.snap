// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@lingui/babel-plugin-extract-messages CallExpression i18n._() should extract messages from i18n._ call expressions 1`] = `
[
  {
    comment: undefined,
    context: undefined,
    id: Message,
    message: undefined,
    origin: [
      js-call-expression.js,
      1,
    ],
  },
  {
    comment: description,
    context: undefined,
    id: Description,
    message: undefined,
    origin: [
      js-call-expression.js,
      3,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: ID,
    message: Message with id,
    origin: [
      js-call-expression.js,
      5,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: Values {param},
    message: undefined,
    origin: [
      js-call-expression.js,
      7,
    ],
  },
  {
    comment: undefined,
    context: Context1,
    id: Some id,
    message: undefined,
    origin: [
      js-call-expression.js,
      9,
    ],
  },
  {
    comment: My comment,
    context: undefined,
    id: my.id,
    message: My Id Message,
    origin: [
      js-call-expression.js,
      12,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: Aliased Message,
    message: undefined,
    origin: [
      js-call-expression.js,
      19,
    ],
  },
  {
    comment: My comment,
    context: undefined,
    id: my.id,
    message: My Id Message,
    origin: [
      js-call-expression.js,
      22,
    ],
  },
]
`;

exports[`@lingui/babel-plugin-extract-messages MessageDescriptor should extract messages from MessageDescriptors 1`] = `
[
  {
    comment: undefined,
    context: undefined,
    id: Message,
    message: undefined,
    origin: [
      js-message-descriptor.js,
      1,
    ],
  },
  {
    comment: description,
    context: undefined,
    id: Description,
    message: undefined,
    origin: [
      js-message-descriptor.js,
      3,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: ID,
    message: Message with id,
    origin: [
      js-message-descriptor.js,
      5,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: Values {param},
    message: undefined,
    origin: [
      js-message-descriptor.js,
      7,
    ],
  },
  {
    comment: undefined,
    context: Context1,
    id: Some id,
    message: undefined,
    origin: [
      js-message-descriptor.js,
      9,
    ],
  },
]
`;

exports[`@lingui/babel-plugin-extract-messages should extract Plural messages from JSX files when there's no Trans tag (integration) 1`] = `
[
  {
    comment: undefined,
    context: undefined,
    id: esnaQO,
    message: {count, plural, one {# book} other {# books}},
    origin: [
      jsx-without-trans.js,
      3,
    ],
  },
  {
    comment: undefined,
    context: Some context,
    id: 8qNz+K,
    message: {count, plural, one {# book} other {# books}},
    origin: [
      jsx-without-trans.js,
      4,
    ],
  },
]
`;

exports[`@lingui/babel-plugin-extract-messages should extract all messages from JS files (macros) 1`] = `
[
  {
    comment: undefined,
    context: undefined,
    id: xDAtGP,
    message: Message,
    origin: [
      js-with-macros.js,
      3,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: xDAtGP,
    message: Message,
    origin: [
      js-with-macros.js,
      5,
    ],
  },
  {
    comment: description,
    context: undefined,
    id: Nu4oKW,
    message: Description,
    origin: [
      js-with-macros.js,
      7,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: ID,
    message: Message with id,
    origin: [
      js-with-macros.js,
      12,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: QCVtWw,
    message: Values {param},
    origin: [
      js-with-macros.js,
      17,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: ID Some,
    message: Message with id some,
    origin: [
      js-with-macros.js,
      19,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: Backtick,
    message: undefined,
    origin: [
      js-with-macros.js,
      24,
    ],
  },
  {
    comment: undefined,
    context: Context1,
    id: Some ID,
    message: undefined,
    origin: [
      js-with-macros.js,
      28,
    ],
  },
  {
    comment: undefined,
    context: Context1,
    id: Some other ID,
    message: undefined,
    origin: [
      js-with-macros.js,
      33,
    ],
  },
  {
    comment: undefined,
    context: Context2,
    id: Some ID,
    message: undefined,
    origin: [
      js-with-macros.js,
      38,
    ],
  },
  {
    comment: undefined,
    context: Context2,
    id: Some ID,
    message: undefined,
    origin: [
      js-with-macros.js,
      43,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: sD7MQ4,
    message: TplLiteral,
    origin: [
      js-with-macros.js,
      48,
    ],
  },
]
`;

exports[`@lingui/babel-plugin-extract-messages should extract all messages from JSX files (macros) 1`] = `
[
  {
    comment: undefined,
    context: undefined,
    id: d1Kdl3,
    message: Hi, my name is {name},
    origin: [
      jsx-with-macros.js,
      3,
    ],
  },
  {
    comment: undefined,
    context: Context1,
    id: YikuIL,
    message: Some message,
    origin: [
      jsx-with-macros.js,
      4,
    ],
  },
  {
    comment: undefined,
    context: Context1,
    id: LBCs5C,
    message: Some other message,
    origin: [
      jsx-with-macros.js,
      5,
    ],
  },
  {
    comment: undefined,
    context: Context2,
    id: ru2rzr,
    message: Some message,
    origin: [
      jsx-with-macros.js,
      6,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: MHrjPM,
    message: Title,
    origin: [
      jsx-with-macros.js,
      7,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: esnaQO,
    message: {count, plural, one {# book} other {# books}},
    origin: [
      jsx-with-macros.js,
      9,
    ],
  },
]
`;

exports[`@lingui/babel-plugin-extract-messages should extract all messages from JSX files 1`] = `
[
  {
    comment: Description,
    context: undefined,
    id: msg.hello,
    message: undefined,
    origin: [
      jsx-without-macros.js,
      5,
    ],
  },
  {
    comment: undefined,
    context: Context1,
    id: msg.context,
    message: undefined,
    origin: [
      jsx-without-macros.js,
      6,
    ],
  },
  {
    comment: undefined,
    context: Context1,
    id: msg.notcontext,
    message: undefined,
    origin: [
      jsx-without-macros.js,
      7,
    ],
  },
  {
    comment: undefined,
    context: Context2,
    id: msg.context,
    message: undefined,
    origin: [
      jsx-without-macros.js,
      8,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: msg.default,
    message: Hello World,
    origin: [
      jsx-without-macros.js,
      9,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: msg.default,
    message: Hello World,
    origin: [
      jsx-without-macros.js,
      10,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: Hi, my name is <0>{name}</0>,
    message: undefined,
    origin: [
      jsx-without-macros.js,
      11,
    ],
  },
]
`;
